### PR TITLE
Simplification of GetHttpMethod and httpMethod in WebCmdlets

### PR DESCRIPTION
**Pull request by CarloToso**
# PR Summary

- GetHttpMethod --> switch expression

- Simplification of setting httpMethod

-  Remove IsStandardMethodSet(), IsCustomMethodSet()

## PR Context
Improve readibility of the WebCmdlets, cleanup

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs [hunk 1]
```diff
@@ -20,35 +20,33 @@ namespace Microsoft.PowerShell.Commands
...
         /// <summary>
         /// Gets or sets the parameter Method.
         /// </summary>
-        [Parameter(ParameterSetName = "StandardMethod")]
-        [Parameter(ParameterSetName = "StandardMethodNoProxy")]
+        [Parameter]
         public override WebRequestMethod Method
         {
-            get { return base.Method; }
+            get => base.Method;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs:34]: I'd move such changes in separate PR since I have concerns about parameterset changes.

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs:34]: What are your concerns, how can I address them?

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs:34]: These are functional changes.  It's better to separate them from the style changes. It makes code review easier.

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs]: I separated them

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 2]
```diff
@@ -1221,7 +1183,7 @@ internal virtual void FillRequestStream(HttpRequestMessage request)
             if (request.Content is null)
             {
                 // If this is a Get request and there is no content, then don't fill in the content as empty content gets rejected by some web services per RFC7230
-                if ((IsStandardMethodSet() && request.Method == HttpMethod.Get && ContentType is null) || (IsCustomMethodSet() && CustomMethod.ToUpperInvariant() == "GET"))
+                if ((request.Method == HttpMethod.Get && ContentType is null) || CustomMethod == "GET")
```
Comment (CarloToso) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1186]: I think the logic in this `if` is wrong, I think it should be:
```cs
if ((request.Method == HttpMethod.Get || CustomMethod == "GET") && ContentType is null)
```

**Comment by iSazonov** [comment]
@CarloToso Of course cmdlets with lots of parameters and parameter sets look complicated nevertheless I don't see the benefit of removing these parameter sets.

---

I am somewhat surprised that parameters are declared as virtual and we override them. We can't make them not virtual without PowerShell Committee approval but we could remove the overrides.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs (hunk 2 above)
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: CustomMethod was actually useless

**Comment by CarloToso** [comment]
@iSazonov I'll add back the parameters set, is the rest of the code ok?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 3]
```diff
@@ -262,7 +262,7 @@ public abstract partial class WebRequestPSCmdlet : PSCmdlet
         /// </summary>
         [Parameter(ParameterSetName = "StandardMethod")]
         [Parameter(ParameterSetName = "StandardMethodNoProxy")]
-        public virtual WebRequestMethod Method { get; set; } = WebRequestMethod.Default;
+        public virtual WebRequestMethod Method { get; set; } = WebRequestMethod.Get;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:265]: Since WebRequestMethod is public we cannot change it (remove the Default). So please keep old default value here and below in conditions.

**Comment by iSazonov** [comment]
> I propose we change the ParameterSetName in the Proxy and NoProxy region from [StandardMethod, CustomMethod -->Proxy] and [StandardMethodNoProxy, CustomMethodNoProxy --> NoProxy]

I don't see value in the change. (Also it can break scripts working with meta data.)

**Comment by CarloToso** [comment]
At the moment WebRequestMethod does not contain the [CONNECT](https://httpwg.org/specs/rfc9110.html#CONNECT) Http request method, should we do something about it?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 4]
```diff
@@ -903,25 +899,11 @@ public abstract partial class WebRequestPSCmdlet : PSCmdlet
 
         private HttpMethod GetHttpMethod(WebRequestMethod method)
         {
-            switch (Method)
+            switch (method)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:902]: I wonder why do we need to remove mappings and use default case with allocation?
I don't see value in the change.

Also the method should be static.

**Comment by iSazonov** [comment]
> At the moment WebRequestMethod does not contain the [CONNECT](https://httpwg.org/specs/rfc9110.html#CONNECT) Http request method, should we do something about it?

Please open new issue with investigation why we should add.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs (hunk 4 above)
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:902]: The code looked cleaner without them and we already didn't have mappings for Merge and Patch methods, so I didn't see any harm in removing all the others.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:902]: Although it's not the hot path I always prefer to stick to the simple rule - if we can not do allocations, it's better not to do them. This is especially important for code that will be used by others in scenarios that we don't know.
So I think this change is unwarranted.
As for the missing values, it is better to discuss this in a new issue to see if it is really often used to have them explicitly.

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:902]: Would it be better if I rewrote GetHttpMethod like this and added Patch to the mapping ?

```cs
        private static HttpMethod GetHttpMethod(WebRequestMethod method) => method switch
        {
            WebRequestMethod.Default or
            WebRequestMethod.Get => HttpMethod.Get,
            WebRequestMethod.Head => HttpMethod.Head,
            WebRequestMethod.Post => HttpMethod.Post,
            WebRequestMethod.Put => HttpMethod.Put,
            WebRequestMethod.Delete => HttpMethod.Delete,
            WebRequestMethod.Trace => HttpMethod.Trace,
            WebRequestMethod.Options => HttpMethod.Options,
            WebRequestMethod.Patch => HttpMethod.Patch,
            _ => new HttpMethod(method.ToString().ToUpperInvariant()), //Merge
        };
```

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:902]: Looks good.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 5]
```diff
@@ -897,16 +897,19 @@ public abstract partial class WebRequestPSCmdlet : PSCmdlet
...
+            WebRequestMethod.Default or
+            WebRequestMethod.Get => HttpMethod.Get,
+            WebRequestMethod.Head => HttpMethod.Head,
+            WebRequestMethod.Post => HttpMethod.Post,
+            WebRequestMethod.Put => HttpMethod.Put,
+            WebRequestMethod.Delete => HttpMethod.Delete,
+            WebRequestMethod.Trace => HttpMethod.Trace,
+            WebRequestMethod.Options => HttpMethod.Options,
+            WebRequestMethod.Patch => HttpMethod.Patch,
+            _ => new HttpMethod(method.ToString().ToUpperInvariant()), //Merge
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:911]: I wonder why is the comment needed?

And maybe sort elements by ABC?

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: done

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 6]
```diff
@@ -899,16 +899,15 @@ public abstract partial class WebRequestPSCmdlet : PSCmdlet
...
+            WebRequestMethod.Patch => HttpMethod.Patch,
             WebRequestMethod.Post => HttpMethod.Post,
             WebRequestMethod.Put => HttpMethod.Put,
-            WebRequestMethod.Delete => HttpMethod.Delete,
-            WebRequestMethod.Trace => HttpMethod.Trace,
             WebRequestMethod.Options => HttpMethod.Options,
-            WebRequestMethod.Patch => HttpMethod.Patch,
-            _ => new HttpMethod(method.ToString().ToUpperInvariant()), //Merge
+            WebRequestMethod.Trace => HttpMethod.Trace,
+            _ => new HttpMethod(method.ToString().ToUpperInvariant()),
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:910]: We can remove final comma.

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@CarloToso Please update the PR description.
